pkg/planner/core/logical_plan_builder.go

```diff
@@ -5256,86 +5258,209 @@ func buildColumns2HandleWithWrtiableColumns(
+	deletableCols := t.DeletableCols()
+	deletableIdxs := t.DeletableIndices()
+	publicCols := t.Cols()
 	tblLen := len(deletableCols)
-	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
-	if err != nil {
-		return TblColPosInfo{}, err
+
+	// Fix the start position of the columns.
+	originalStart := colPosInfo.Start
+	colPosInfo.Start -= prePrunedCount
+
+	// Mark the columns in handle.
+	fixedPos := make(map[int]int, len(deletableCols))
+	for i := 0; i < colPosInfo.HandleCols.NumCols(); i++ {
+		col := colPosInfo.HandleCols.GetCol(i)
+		fixedPos[col.Index-originalStart] = 0
 	}
-	end := offset + tblLen
-
-	// Index only records its columns' offsets in the deletableCols(the whole columns).
-	// So we need to first change the offsets to the real position of SELECT's output.
-	offsetMap := make(map[int]int, len(deletableCols))
-	// For multi-delete case, the offset is recorded for the mixed row. We need to use its position in single table's row layout.
-	//	e.g. The multi-delete case is [t1.a, t1.b, t2.a, t2.b] and There's a index [t2.b].
-	//	     The idxCols' original offset is [3]. And we should use [1] instead.
-	for i, name := range names[offset:end] {
-		found := false
-		for j, col := range deletableCols {
-			if col.Name.L == name.ColName.L {
-				offsetMap[j] = i
-				found = true
-				break
+
+	// Mark the columns in indexes.
+	for _, idx := range deletableIdxs {
+		for _, col := range idx.Meta().Columns {
+			if col.Offset+originalStart >= len(names) || deletableCols[col.Offset].Name.L != names[col.Offset+originalStart].ColName.L {
+				return 0, plannererrors.ErrDeleteNotFoundColumn.GenWithStackByArgs(col.Name.O, tblInfo.Name.O)
 			}
+			fixedPos[col.Offset] = 0
 		}
-		if !found {
-			return TblColPosInfo{}, plannererrors.ErrDeleteNotFoundColumn.GenWithStackByArgs(name.ColName.O, tblInfo.Name.O)
+	}
+
+	// Fix the column offsets.
+	pruned := 0
+	for i := 0; i < tblLen; i++ {
+		if _, ok := fixedPos[i]; !ok {
+			nonPrunedSet.Clear(uint(i + originalStart))
+			pruned++
+			continue
+		}
+		fixedPos[i] = i - pruned
+	}
+
+	// Fill in the ColumnSizes.
+	colPosInfo.ColumnsSizeHelper = &table.ColumnsSizeHelper{
+		NotPruned:        bitset.New(uint(len(publicCols))),
+		AvgSizes:         make([]float64, 0, len(publicCols)),
+		PublicColsLayout: make([]int, 0, len(publicCols)),
+	}
+	colPosInfo.ColumnsSizeHelper.NotPruned.SetAll()
+	for i, col := range publicCols {
+		// If the column is not pruned, we can use the column data to get a more accurate size.
+		// We just need to record its position info.
+		if _, ok := fixedPos[col.Offset]; ok {
+			colPosInfo.ColumnsSizeHelper.PublicColsLayout = append(colPosInfo.ColumnsSizeHelper.PublicColsLayout, fixedPos[col.Offset])
+			continue
 		}
+		// Otherwise we need to get the average size of the column by its field type.
+		// TODO: use statistics to get a maybe more accurate size.
+		colPosInfo.ColumnsSizeHelper.NotPruned.Clear(uint(i))
```

Comment:
The above comment is strange for this line, how about move this line after next line?